openshift-knative-operator/pkg/common/api.go

```diff
@@ -0,0 +1,177 @@
+			SeccompProfile:           &corev1.SeccompProfile{Type: corev1.SeccompProfileTypeRuntimeDefault},
+		}
+	} else {
+		if container.SecurityContext.RunAsNonRoot == nil {
+			container.SecurityContext.RunAsNonRoot = ptr.Bool(true)
+		}
+		if container.SecurityContext.ReadOnlyRootFilesystem == nil {
+			container.SecurityContext.ReadOnlyRootFilesystem = ptr.Bool(true)
+		}
+		if container.SecurityContext.AllowPrivilegeEscalation == nil {
+			container.SecurityContext.AllowPrivilegeEscalation = ptr.Bool(false)
+		}
+		container.SecurityContext.Capabilities = &corev1.Capabilities{Drop: []corev1.Capability{"ALL"}}
+		if container.SecurityContext.SeccompProfile == nil {
+			container.SecurityContext.SeccompProfile = &corev1.SeccompProfile{Type: corev1.SeccompProfileTypeRuntimeDefault}
+		}
+	}
+}
+
+// CheckMinimumVersion checks if current K8s version we are on is higher than the one passed.
+// If an error is returned then we
+func CheckMinimumVersion(versioner discovery.ServerVersionInterface, version string) error {
+	v, err := versioner.ServerVersion()
+	if err != nil {
+		return err
+	}
+	currentVersion, err := semver.Make(normalizeVersion(v.GitVersion))
+	if err != nil {
+		return err
+	}
+
+	minimumVersion, err := semver.Make(normalizeVersion(version))
+	if err != nil {
+		return err
+	}
+
+	// If no specific pre-release requirement is set, we default to "-0" to always allow
+	// pre-release versions of the same Major.Minor.Patch version.
+	if len(minimumVersion.Pre) == 0 {
+		minimumVersion.Pre = []semver.PRVersion{{VersionNum: 0, IsNum: true}}
+	}
+
+	if currentVersion.LT(minimumVersion) {
+		return fmt.Errorf("kubernetes version %q is not compatible, need at least %q",
+			currentVersion, minimumVersion)
+	}
+	return nil
+}
+
+// DeprecatedAPIsTranformersFromConfig check if we are on the right K8s version and return
+// the related transformers. Meant to be used by the knative-openshift operator which uses controller runtime
+// and for which we need to construct the discovery value.
+func DeprecatedAPIsTranformersFromConfig() []mf.Transformer {
+	cfg, err := config.GetConfig()
+	if err != nil {
+		panic(err)
+	}
+	clset := kubernetes.NewForConfigOrDie(cfg)
+	return DeprecatedAPIsTranformers(clset.Discovery())
+}
+
+// DeprecatedAPIsTranformers check if we are on the right K8s version and return
+// the related transformers.
+func DeprecatedAPIsTranformers(d discovery.DiscoveryInterface) []mf.Transformer {
+	transformers := []mf.Transformer{}
+	// Enforce the new version, try to upgrade existing resources for 4.11 to also avoid warnings.
+	// The policy/v1beta1 API version of PodDisruptionBudget will no longer be served in v1.25.
+	// The autoscaling/v2beta2 API version of HorizontalPodAutoscaler will no longer be served in v1.26
+	// TODO: When we move away from releases that bring v1beta1 we can remove this part
+	if err := CheckMinimumVersion(d, "1.24.0"); err == nil {
```

Comment:
Picking 1.24 instead of 1.25 here to avoid warnings 4.11 and a scenario where restricted policy is enforced.